### PR TITLE
sw_engine: code refactoring.

### PR DESCRIPTION
**Pull request by hermet**
simplify neon code and fix the wrong range access.

### CODE REVIEW

File: src/lib/sw_engine/tvgSwRasterNeon.h [hunk 1]
```diff
@@ -91,36 +91,32 @@ static inline bool neonRasterTranslucentRle(SwSurface* surface, const SwRleData*
...
-        uint32_t iterations = w / 2;
-        uint32_t left = w % 2;
-
-        if (align) *dst = color + ALPHA_BLEND(*dst, ialpha);
-
-        for (uint32_t x = 0; x < iterations; ++x) vDst[x] = vadd_u8(vColor, ALPHA_BLEND_NEON(vDst[x], vIalpha));
+        for (uint32_t x = 0; x <  w / 2; ++x)
+            vDst[x] = vadd_u8(vColor, ALPHA_BLEND_NEON(vDst[x], vIalpha));
         
-        if (left) dst[w] = color + ALPHA_BLEND(dst[w], ialpha);
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRasterNeon.h:123]: dst[w] is out of boundary.

File: src/lib/sw_engine/tvgSwRasterNeon.h [hunk 2]
```diff
@@ -91,36 +91,32 @@ static inline bool neonRasterTranslucentRle(SwSurface* surface, const SwRleData*
...
-    uint8x8_t *vDst = NULL;
+    auto vColor = static_cast<uint8x8_t>(vdup_n_u32(color));
+    auto vIalpha = static_cast<uint8x8_t>(vdup_n_u8((uint8_t) ialpha));
+    uint8x8_t* vDst = nullptr;
 
     for (uint32_t y = 0; y < h; ++y) {
-        uint32_t align = 0;
-        uint32_t *dst = &buffer[y * surface->stride];
+
+        auto *dst = &buffer[y * surface->stride];
```
Comment (mgrudzinska) [inline src/lib/sw_engine/tvgSwRasterNeon.h:105]: -*

File: src/lib/sw_engine/tvgSwRasterNeon.h [hunk 3]
```diff
@@ -91,36 +91,32 @@ static inline bool neonRasterTranslucentRle(SwSurface* surface, const SwRleData*
...
 
         if ((((uint32_t) dst) & 0x7) != 0) {
-            vDst = (uint8x8_t*)(dst+1);
-            align = 1;
+            //fill not aligned byte
+            *dst = color + ALPHA_BLEND(*dst, ialpha);
+            vDst = (uint8x8_t*) (dst + 1);
```
Comment (mgrudzinska) [inline src/lib/sw_engine/tvgSwRasterNeon.h:110]: just a comment - in some places you've changed the c-style casting to static_cast and in some you didn't.

File: src/lib/sw_engine/tvgSwRasterNeon.h [hunk 4]
```diff
@@ -91,36 +91,32 @@ static inline bool neonRasterTranslucentRle(SwSurface* surface, const SwRleData*
...
             vDst = (uint8x8_t*) dst;
         }
 
-        uint32_t iterations = w / 2;
-        uint32_t left = w % 2;
-
-        if (align) *dst = color + ALPHA_BLEND(*dst, ialpha);
-
-        for (uint32_t x = 0; x < iterations; ++x) vDst[x] = vadd_u8(vColor, ALPHA_BLEND_NEON(vDst[x], vIalpha));
+        for (uint32_t x = 0; x <  w / 2; ++x)
```
Comment (mgrudzinska) [inline src/lib/sw_engine/tvgSwRasterNeon.h:115]: there is a problem in the case where `w` is even and the `dst` address is not aligned.
lets take `w=4`:
- first we fill the not aligned bytes (line 109) - 1 pixel
- in the for loop (line 115) `x` can be 0 and 1 so we try to fill another 4 pixels, but there were only 4 in total

**Review by mgrudzinska — CHANGES_REQUESTED** [review]

File: src/lib/sw_engine/tvgSwRasterNeon.h (hunk 3 above)
Comment (hermet) [reply src/lib/sw_engine/tvgSwRasterNeon.h:110]: Both are allowed, not forced to use one way.

File: src/lib/sw_engine/tvgSwRasterNeon.h (hunk 2 above)
Comment (hermet) [reply src/lib/sw_engine/tvgSwRasterNeon.h:105]: fixed.

File: src/lib/sw_engine/tvgSwRasterNeon.h (hunk 4 above)
Comment (hermet) [reply src/lib/sw_engine/tvgSwRasterNeon.h:115]: I think that is another story out of this patch boundary. Please make a patch.

Comment (mgrudzinska) [reply src/lib/sw_engine/tvgSwRasterNeon.h:115]: I opened #747.
If you first merge your, I'll solve conflicts in mine

**Comment by hermet** [comment]
@mgrudzinska changed bool align -> uint16_t

**Comment by mgrudzinska** [comment]
@hermet please rebase and merge

**Review by mgrudzinska — APPROVED** [review]

**Pull request merged**